### PR TITLE
CodeSearch: Switch to new API

### LINKED ISSUE #3011: Search Code: Newlines treated literally
When searching for https://duckduckgo.com/?q=perl+%3C&ia=answer the `Search Code` Instant Answer triggers, but the output is different to what it should be.

Shown result:

``` perl
# Hello world in perl

print "Hello World!
";
```

Expected:

``` perl
# Hello world in perl

print "Hello World!\n";
```

It looks like the newline is being treated literally.

(cc @sahildua2305)

---

IA Page:  http://duck.co/ia/view/search_code 
[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @sahildua2305

### PR DESCRIPTION
**Pull request by jrasanen**
###### What does your Pull Request do (check all that apply)?

Choose the most relevant items and use the following title template to name
your Pull Request.
- [ ] New Instant Answer: **`New New  Instant Answer`**
- [X] Improvement
  - [ ] Bug fix: **`{IA Name}: Fix {Issue number or one-line description}`**
  - [X] Enhancement: **`Code search: New uses sourcegraph`**
- [ ] Non–Instant Answer
  - [ ] Other (Role, Template, Test, Documentation, etc.): **`{SpiceRole/Templates/Tests/Docs}: {Short Description}`**
###### Description of changes

This pull request is work in progress, using sourcegraph for new code search.

Fixes #3011 
###### Related Issues or Discussions

Related to [Improvement to Python-search](https://forum.duckduckhack.com/t/python-search-overview/97)

---

Instant Answer Page: https://duck.co/ia/view/code_search

[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @mention

### CODE REVIEW

**Comment by daxtheduck** [comment]
@jrasanen Thanks for the pull request! Unfortunately, we're only accepting contributions and improvements to a limited portion of the search space.
            
The staff and community are currently focused on making DuckDuckGo the [best search engine for programmers](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53). To ensure we're making as much progress as possible, **we are only accepting Instant Answers and updates related to the programming mission**.
    
If this is not related to the Programming Mission, it will likely be closed but you can:
    
1. Help us prevent further confusion by telling us where you found out about DuckDuckHack or where we could've alerted you earlier.
2. Keep this code handy! We'll be expanding into other search topics in the future!
3. Join us at our new home: https://forum.duckduckhack.com/

If this is related to the Programming Mission, it will be reviewed by staff and the community as soon as possible!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by moollaza** [comment]
@jrasanen thanks for this! It looks like CodeSearch.t is failing though -- can you please update the tests to pass? Once that's done I can deploy this to Beta 👍

**Comment by moollaza** [comment]
@jrasanen Ping! I'm just checking in to see if you're still interested in finishing this one up? We'd love to see this get onto Beta!

File: lib/DDG/Spice/CodeSearch.pm [hunk 1]
```diff
@@ -6,7 +6,8 @@ use DDG::Spice;
 
 triggers startend => "code", "example";
 
-spice to => 'https://searchcode.com/api/jsonp_codesearch_I/?q=$1&callback={{callback}}';
+spice to => 'https://sourcegraph.com/.api/global-search?Query=$1&Limit=1&Fast=0';
+spice wrap_jsonp_callback => 1;
 
 my $languages = join "|", share('languages.txt')->slurp;
```
Comment (moollaza) [inline lib/DDG/Spice/CodeSearch.pm]: @jrasanen let's tighten this up. For now, let's switch the trigger to `python`. We can relax the triggering once this is live as necessary

File: lib/DDG/Spice/CodeSearch.pm [hunk 2]
```diff
@@ -19,7 +20,7 @@ handle remainder => sub {
 
         s/\s*$match\s*//i;
 
-        return "lang:".$match." ".$_;
+        return $match." ".$_;
```
Comment (moollaza) [inline lib/DDG/Spice/CodeSearch.pm]: is there no better way of specifying the language we want results for? when I played around with the API I was noticing results from several languages.

Do we have access to any API documentation?

File: share/spice/code_search/code_search.js [hunk 3]
```diff
@@ -2,37 +2,27 @@
...
+                    name: item.Data.Name,
+                    info: item.DocHTML.__html
                 };
             },
             meta: {
-                sourceUrl: 'https://searchcode.com/?q=' + query + '&cs=true',
-                sourceIconUrl: 'https://searchcode.com/static/favicon.ico',
-                sourceName: 'searchcode'
+                sourceUrl: 'https://' + query,
+                sourceIconUrl: 'https://assets-cdn.github.com/favicon.ico',
```
Comment (moollaza) [inline share/spice/code_search/code_search.js]: You should be able to remove this  -- we'll get the favicon from the `moreAt` link

Comment (moollaza) [inline share/spice/code_search/code_search.js]: Also, I don't think every result goes to GitHub?

Actually, the more at link should probably go to a search on SourceGraph. We can create additional links (e.g. a clickable title, or subtitle links) to view the actual code

**Comment by moollaza** [comment]
@jrasanen we still need to update the tests as well to pass and then I can deploy this PR to Beta, along with other updates 👍

File: share/spice/code_search/code_search.js (hunk 3 above)
Comment (jrasanen) [inline share/spice/code_search/code_search.js]: Sourcegraph does have this filebrowser, https://sourcegraph.com/github.com/newvem/beautifulsoup4/-/blob/bs4/element.py

**Comment by moollaza** [comment]
DuckDuckHack is now in Maintenance Mode and from now on, we are only accepting issues and PRs for essential bugs and bug fixes.

Unfortunately, with the above in mind, this isn't something we can action and will be closed.

We appreciate you taking the time to contribute and apologize for not being able to triage this issue.

**Pull request closed without merging**